classes/product/SpecificPriceFormatter.php

```diff
@@ -0,0 +1,179 @@
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+use PrestaShop\PrestaShop\Adapter\Product\PriceFormatter;
+
+/**
+ * This class is an extract of the ProductController->formatQuantityDiscounts() method
+ * It is still legacy code and need some heavy refactoring.
+ * This code has been extracted here for unit testing purpose
+ *
+ * @todo Refactor this class
+ */
+class SpecificPriceFormatterCore
+{
+    /**
+     * Calculation method to be used (tax included or not?)
+     *
+     * @var bool
+     */
+    private $isTaxIncluded;
+
+    /**
+     * Specific price data array
+     *
+     * @var array
+     */
+    private $specificPrice;
+
+    /**
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * SpecificPriceFormatter constructor.
+     *
+     * @param array $specificPrice
+     * @param bool $isTaxIncluded
+     * @param Context $context
+     */
+    public function __construct(array $specificPrice, bool $isTaxIncluded, Context $context)
+    {
+        $this->specificPrice = $specificPrice;
+        $this->isTaxIncluded = $isTaxIncluded;
+        $this->context = $context;
+    }
+
+    /**
+     * This is legacy code extracted from ProductController and it should be refactored
+     *
+     * @param float $initialPrice
+     * @param float $tax_rate
+     * @param float $ecotax_amount
+     *
+     * @return array
+     */
+    public function formatSpecificPrice($initialPrice, $tax_rate, $ecotax_amount)
+    {
+        $priceFormatter = new PriceFormatter();
+
+        $this->specificPrice['quantity'] = &$this->specificPrice['from_quantity'];
+        if ($this->specificPrice['price'] >= 0) {
+            // The price may be directly set
+
+            /** @var float $currentPriceDefaultCurrency current price with taxes in default currency */
+            $currentPriceDefaultCurrency = $this->calculateSpecificPrice(
+                $this->specificPrice['price'],
+                (bool) $this->specificPrice['reduction_tax'],
+                $tax_rate,
+                $ecotax_amount
+            );
+
+            // Since this price is set in default currency,
+            // we need to convert it into current currency
+            $this->specificPrice['id_currency'];
+            $currentPriceCurrentCurrency = \Tools::convertPrice($currentPriceDefaultCurrency, $this->context->currency, true, $this->context);
```

Comment:
Seems like you only need the context to use this method
But if you look at the `Tools::convertPrice` it only needs the context as a fallback if the currency parameter is not provided
So instead of injecting the context you could simply inject the currency or use it as a parameter of your method, this way you could call `\Tools::convertPrice($currentPriceDefaultCurrency, $this->currency, true);` with the same result but then you can remove the dependency to Context

Reply:
@jolelievre  done !